bindings/python/setup.py

```diff
@@ -1,195 +1,219 @@
+    # request, so that <define>TORRENT_FOO becomes something like
+    # <define>TORRENT_FOO,<python>3.7,<target-os>linux:... which causes chaos.
+    # We should always define a custom feature for the condition.
+    write("import feature ;\n")
+    write("feature.feature libtorrent-python : on ;\n")
+
+    # python.jam tries to determine correct include and library paths. Per
+    # https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=691378 , include
+    # detection is broken, but debian's fix is also broken (invokes a global
+    # pythonX.Y instead of the passed interpreter)
+    paths = sysconfig.get_paths()
+    includes = [paths["include"], paths["platinclude"]]
+
+    write("using python")
+    write(f" : {sysconfig.get_python_version()}")
+    write(f' : "{sys.executable}"')
+    write(" : ")
+    write(" ".join(f'"{path}"' for path in includes))
+    write(" :")  # libraries
+    write(" : <libtorrent-python>on")
+    # Note that on debian, the extension suffix is overwritten, but it's
+    # necessary everywhere else, or else b2 will just build "libtorrent.so".
+    # python.jam appends SHLIB_SUFFIX on its own.
+    ext_suffix = sysconfig.get_config_var("EXT_SUFFIX")
+    shlib_suffix = sysconfig.get_config_var("SHLIB_SUFFIX")
+    ext_suffix = ext_suffix[: -len(shlib_suffix)]
+    write(f' : "{ext_suffix}"')
+    write(" ;\n")
+
+
+BuildExtBase = _build_ext_lib.build_ext
+
+
+class LibtorrentBuildExt(BuildExtBase):
+
+    user_options = BuildExtBase.user_options + [
+        (
+            "libtorrent-link=",
+            None,
+            "how to link to libtorrent ('static' or 'shared')",
+        ),
+        (
+            "boost-link=",
+            None,
+            "how to link to boost-python ('static' or 'shared')",
+        ),
+        ("toolset=", None, "b2 toolset"),
+        ("pic", None, "whether to compile with -fPIC"),
+        ("optimization=", None, "b2 optimization mode"),
+        (
+            "hash",
+            None,
+            "use a property hash for the build directory, rather than "
+            "property subdirectories",
+        ),
+        ("cxxstd=", None, "boost cxxstd value (11, 14, 17, etc)"),
     ]
 
-setup(
-    name='python-libtorrent',
-    version='1.2.11',
-    author='Arvid Norberg',
-    author_email='arvid@libtorrent.org',
-    description='Python bindings for libtorrent-rasterbar',
-    long_description='Python bindings for libtorrent-rasterbar',
-    url='http://libtorrent.org',
-    platforms=[platform.system() + '-' + platform.machine()],
-    license='BSD',
-    packages=packages,
-    ext_modules=ext
+    boolean_options = BuildExtBase.boolean_options + ["pic", "hash"]
+
+    def initialize_options(self):
+        self.libtorrent_link = None
+        self.boost_link = None
+        self.toolset = None
+        self.pic = None
+        self.optimization = None
+        self.hash = None
+        if platform.system() == "Darwin":
```

Comment:
I don't think this distinction is important. You just happen to have an older compiler on MacOS that doesn't default to C++11. The more interesting question is, which ABI is python itself use? perhaps C++11 is too old.

Reply:
> I don't think this distinction is important. You just happen to have an older compiler on MacOS that doesn't default to C++11.

I see. Do you have opinions on whether it would be better to default `toolset=gcc` on darwin, rather than default `cxxstd=11`? I'll play with this.

I guess what's really needed is a default of `cxxstd=11` if `toolset=darwin`, but `setup.py` doesn't know if `toolset=darwin` will be autodiscovered in b2. IMO would be much better to just apply this in some `Jamfile` in RC_1_2, rather than here.

> The more interesting question is, which ABI is python itself use? perhaps C++11 is too old.

cpython's ABI is C only. So `cxxstd`'s ABI just matters for the extension's downstream dynamic linkage. Ideally, when packaging a wheel for pypi, the choice of `cxxstd` should never matter to consumers.

(There are various C ABIs defined, but `setup.py`'s scope is to only know about the ABI of the running interpreter, and configure boost to build for that)

Reply:
> IMO would be much better to just apply this in some Jamfile in RC_1_2, rather than here.

It is in RC_2_0 and later

> cpython's ABI is C only. So cxxstd's ABI just matters for the extension's downstream dynamic linkage.

Which C++ standard is used affects which C standard the compiler is using as well. The C standard doesn't define an ABI, just like C++ doesn't. It may be easier for platforms to maintain a stable ABI in C, but there's no guarantee. windows/MSVC is a good example.

I imagine specifying `stdcxx=11` here is safe, but with MSVC, the version of the compiler depends on the version of python.

Reply:
In truth I don't think we need to work too hard to solve this here.

I don't think invoking `setup.py` with no args needs to produce _the_ perfect most-compatible build. When packaging wheels for pypi, I expect we'd do things like `--libtorrent-link=static --boost-link=static` which wouldn't be appropriate in most other cases.

I just wanted `setup.py` with no args to produce _something_ useful, both to help users, and to trim down os-specific overrides in CI. I expect the impact of changing this default later will be low.

Reply:
For what it's worth, I haven't seen evidence that python is required to be compiled with any particular C standard on mac/linux.

On linux particularly, I struggle to think of any problems that would be caused for an executable linking to a library, if they differed in C standard alone. I would expect versioned symbols in libc to resolve the differences (in a perfect world). I don't have much macos experience so I don't know if they have something similar.